File: 2381.Shifting-Letters-II/2381.shifting-letters-ii.py
#
# @lc app=leetcode id=2381 lang=python
#
# [2381] Shifting Letters II
#

# @lc code=start
class Solution(object):
    def shiftingLetters(self, s, shifts):
        temp_s = list(s)
        shift = [0]*(len(temp_s)+1)
        ans = ""
        for i in shifts:
            if i[2] == 1: # forward
                shift[i[0]] += 1
                shift[i[1]+1] -= 1
            elif i[2] == 0 : #backward
                shift[i[0]] -= 1
                shift[i[1]+1] += 1
        num = 0 
        buf = 0 
        for i in range(len(shift)-1):
            buf += shift[i]
            buf %= 26
            num = ord(temp_s[i])+buf
            if num<97:
                num = 122-(96-num)
            elif num>122:
                num = num-123+97
            ans += chr(num)
        if(len(ans)<len(s)):
            ans += s[3:]
        print(ans)
        return str(ans)
        """
        :type s: str
        :type shifts: List[List[int]]
        :rtype: str
        """
        
# @lc code=end

test = Solution()
s = "abc"
shifts =[[0,1,0],[1,2,1],[0,2,1]]
test.shiftingLetters(s,shifts)

File: 2381.Shifting-Letters-II/2381.shifting-letters-ii.cpp
/*
 * @lc app=leetcode id=2381 lang=cpp
 *
 * [2381] Shifting Letters II
 */

// @lc code=start
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        string ans ="";
        vector<int> shift(s.length()+1,0);
        for(auto it = shifts.begin(); it!= shifts.end();++it){
            if((*it)[2]==1){
                shift[(*it)[0]]+=1;
                shift[(*it)[1]+1]-=1;
            }
            else if((*it)[2]==0){
                shift[(*it)[0]]-=1;
                shift[(*it)[1]+1]+=1;
            }
        }
        int buf = 0;
        int new_word = 0;
        int num = 0;
        for(auto it= shift.begin();it!=shift.end();++it){
            buf += *it;
            buf %= 26;
            new_word = s[num]+buf;
            if(new_word<97){
                new_word = 122-(96-new_word);
            }
            else if(new_word>122){
                new_word = new_word-123+97;
            }
            ans += char(new_word);
            num += 1;
        }
        return ans.erase(ans.size()-1);
    }
};
// @lc code=end

